File: app/src/main/java/org/wikipedia/appshortcuts/AppShortcuts.java
package org.wikipedia.appshortcuts;

import android.annotation.TargetApi;
import android.content.Intent;
import android.content.pm.ShortcutInfo;
import android.content.pm.ShortcutManager;
import android.graphics.drawable.Icon;
import android.net.Uri;
import android.os.Build;

import org.wikipedia.Constants;
import org.wikipedia.R;
import org.wikipedia.WikipediaApp;
import org.wikipedia.main.MainActivity;
import org.wikipedia.search.SearchActivity;
import org.wikipedia.search.SearchInvokeSource;

import java.util.Arrays;

public class AppShortcuts {

    private static final String ACTION_APP_SHORTCUT = "org.wikipedia.app_shortcut";
    private WikipediaApp app;

    public AppShortcuts() {
        app = WikipediaApp.getInstance();
    }

    @TargetApi(android.os.Build.VERSION_CODES.N_MR1)
    public void init() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
            ShortcutManager shortcutManager = app.getSystemService(ShortcutManager.class);
            shortcutManager.setDynamicShortcuts(Arrays.asList(searchShortcut(), continueReadingShortcut(), randomShortcut()));
        }
    }

    @TargetApi(android.os.Build.VERSION_CODES.N_MR1)
    private ShortcutInfo randomShortcut() {

        ShortcutInfo shortcut = new ShortcutInfo.Builder(app, app.getString(R.string.app_shortcuts_random))
                .setShortLabel(app.getString(R.string.app_shortcuts_random))
                .setLongLabel(app.getString(R.string.app_shortcuts_random))
                .setIcon(Icon.createWithResource(app, R.drawable.appshortcut_ic_random))
                .setIntent(
                        new Intent(Intent.ACTION_MAIN, Uri.EMPTY, app, MainActivity.class)
                                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                                .putExtra(Constants.INTENT_APP_SHORTCUT_RANDOM, true))
                .build();

        return shortcut;
    }

    @TargetApi(android.os.Build.VERSION_CODES.N_MR1)
    private ShortcutInfo continueReadingShortcut() {

        ShortcutInfo shortcut = new ShortcutInfo.Builder(app, app.getString(R.string.app_shortcuts_continue_reading))
                .setShortLabel(app.getString(R.string.app_shortcuts_continue_reading))
                .setLongLabel(app.getString(R.string.app_shortcuts_continue_reading))
                .setIcon(Icon.createWithResource(app, R.drawable.appshortcut_ic_continue_reading))
                .setIntent(new Intent(ACTION_APP_SHORTCUT).putExtra(Constants.INTENT_APP_SHORTCUT_CONTINUE_READING, true))
                .build();

        return shortcut;
    }

    @TargetApi(android.os.Build.VERSION_CODES.N_MR1)
    private ShortcutInfo searchShortcut() {

        ShortcutInfo shortcut = new ShortcutInfo.Builder(app, app.getString(R.string.app_shortcuts_search))
                .setShortLabel(app.getString(R.string.app_shortcuts_search))
                .setLongLabel(app.getString(R.string.app_shortcuts_search))
                .setIcon(Icon.createWithResource(app, R.drawable.appshortcut_ic_search))
                .setIntent(SearchActivity.newIntent(app, SearchInvokeSource.APP_SHORTCUTS.code(), null).setAction(ACTION_APP_SHORTCUT))
                .build();

        return shortcut;
    }
}

File: app/src/main/java/org/wikipedia/appshortcuts/AppShortcuts.kt
package org.wikipedia.appshortcuts

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.pm.ShortcutInfoCompat
import androidx.core.content.pm.ShortcutManagerCompat
import androidx.core.graphics.drawable.IconCompat
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.wikipedia.Constants
import org.wikipedia.R
import org.wikipedia.main.MainActivity
import org.wikipedia.util.log.L

class AppShortcuts {
    companion object {
        const val ACTION_APP_SHORTCUT = "ACTION_APP_SHORTCUT"
        const val APP_SHORTCUT_ID = "APP_SHORTCUT_ID"
        private const val APP_SHORTCUT_ID_CONTINUE_READING = "shortcutContinueReading"
        private const val APP_SHORTCUT_ID_RANDOM = "shortcutRandom"
        private const val APP_SHORTCUT_ID_SEARCH = "shortcutSearch"
        private const val APP_SHORTCUT_ID_PLACES = "shortcutPlaces"

        fun setShortcuts(app: Context) {
            CoroutineScope(Dispatchers.Default).launch(CoroutineExceptionHandler { _, msg -> run { L.e(msg) } }) {
                val list = listOf(searchShortcut(app), continueReadingShortcut(app), randomShortcut(app), placesShortcut(app))
                if (ShortcutManagerCompat.getDynamicShortcuts(app).map { it.id }.containsAll(list.map { it.id }).not()) {
                    ShortcutManagerCompat.setDynamicShortcuts(app, list)
                } else {
                    L.d("Create dynamic shortcuts skipped.")
                }
            }
        }

        private fun searchShortcut(app: Context): ShortcutInfoCompat {
            return ShortcutInfoCompat.Builder(app, APP_SHORTCUT_ID_SEARCH)
                    .setShortLabel(app.getString(R.string.app_shortcuts_search))
                    .setLongLabel(app.getString(R.string.app_shortcuts_search))
                    .setIcon(IconCompat.createWithResource(app, R.drawable.appshortcut_ic_search))
                    .setIntent(
                            Intent(ACTION_APP_SHORTCUT, Uri.EMPTY, app, MainActivity::class.java)
                                    .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                                    .putExtra(APP_SHORTCUT_ID, APP_SHORTCUT_ID_SEARCH)
                                    .putExtra(Constants.INTENT_APP_SHORTCUT_SEARCH, true))
                    .build()
        }

        private fun randomShortcut(app: Context): ShortcutInfoCompat {
            return ShortcutInfoCompat.Builder(app, APP_SHORTCUT_ID_RANDOM)
                    .setShortLabel(app.getString(R.string.app_shortcuts_random))
                    .setLongLabel(app.getString(R.string.app_shortcuts_random))
                    .setIcon(IconCompat.createWithResource(app, R.drawable.appshortcut_ic_random))
                    .setIntent(
                            Intent(ACTION_APP_SHORTCUT, Uri.EMPTY, app, MainActivity::class.java)
                                    .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                                    .putExtra(APP_SHORTCUT_ID, APP_SHORTCUT_ID_RANDOM)
                                    .putExtra(Constants.INTENT_APP_SHORTCUT_RANDOMIZER, true))
                    .build()
        }

        private fun continueReadingShortcut(app: Context): ShortcutInfoCompat {
            return ShortcutInfoCompat.Builder(app, APP_SHORTCUT_ID_CONTINUE_READING)
                    .setShortLabel(app.getString(R.string.app_shortcuts_continue_reading))
                    .setLongLabel(app.getString(R.string.app_shortcuts_continue_reading))
                    .setIcon(IconCompat.createWithResource(app, R.drawable.appshortcut_ic_continue_reading))
                    .setIntent(
                            Intent(ACTION_APP_SHORTCUT, Uri.EMPTY, app, MainActivity::class.java)
                                    .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                                    .putExtra(APP_SHORTCUT_ID, APP_SHORTCUT_ID_CONTINUE_READING)
                                    .putExtra(Constants.INTENT_APP_SHORTCUT_CONTINUE_READING, true))
                    .build()
        }

        private fun placesShortcut(app: Context): ShortcutInfoCompat {
            return ShortcutInfoCompat.Builder(app, APP_SHORTCUT_ID_PLACES)
                .setShortLabel(app.getString(R.string.app_shortcuts_places))
                .setLongLabel(app.getString(R.string.app_shortcuts_places))
                .setIcon(IconCompat.createWithResource(app, R.drawable.appshortcut_ic_places))
                .setIntent(
                    Intent(ACTION_APP_SHORTCUT, Uri.EMPTY, app, MainActivity::class.java)
                        .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                        .putExtra(APP_SHORTCUT_ID, APP_SHORTCUT_ID_PLACES)
                        .putExtra(Constants.INTENT_APP_SHORTCUT_PLACES, true))
                .build()
        }
    }
}
